drivers/ng_at86rf2xx/ng_at86rf2xx.c

```diff
@@ -0,0 +1,245 @@
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ng_at86rf2xx
+ * @{
+ *
+ * @file
+ * @brief       Implementation of public functions for AT86RF2xx drivers
+ *
+ * @author      Alaeddine Weslati <alaeddine.weslati@inria.fr>
+ * @author      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include "hwtimer.h"
+#include "periph/cpuid.h"
+#include "net/ng_ieee802154.h"
+#include "net/ng_netbase.h"
+#include "ng_at86rf2xx_registers.h"
+#include "ng_at86rf2xx_internal.h"
+#include "ng_at86rf2xx_netdev.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+
+#define RESET_DELAY             (0U)        /* must be > 625ns */
+
+
+static void _irq_handler(void *arg)
+{
+    msg_t msg;
+    ng_at86rf2xx_t *dev = (ng_at86rf2xx_t *) arg;
+
+    /* tell driver thread about the interrupt */
+    msg.type = NG_NETDEV_MSG_TYPE_EVENT;
+    msg_send(&msg, dev->mac_pid);
+}
+
+int ng_at86rf2xx_init(ng_at86rf2xx_t *dev, spi_t spi, spi_speed_t spi_speed,
+                      gpio_t cs_pin, gpio_t int_pin,
+                      gpio_t sleep_pin, gpio_t reset_pin)
+{
+    dev->driver = &ng_at86rf2xx_driver;
+
+    /* initialize device descriptor */
+    dev->spi = spi;
+    dev->cs_pin = cs_pin;
+    dev->int_pin = int_pin;
+    dev->sleep_pin = sleep_pin;
+    dev->reset_pin = reset_pin;
+
+    /* initialise SPI */
+    spi_init_master(dev->spi, SPI_CONF_FIRST_RISING, spi_speed);
+    /* initialise GPIOs */
+    gpio_init_out(dev->cs_pin, GPIO_NOPULL);
+    gpio_set(dev->cs_pin);
+    gpio_init_out(dev->sleep_pin, GPIO_NOPULL);
+    gpio_clear(dev->sleep_pin);
+    gpio_init_out(dev->reset_pin, GPIO_NOPULL);
+    gpio_set(dev->reset_pin);
+    gpio_init_int(dev->int_pin, GPIO_NOPULL, GPIO_RISING, _irq_handler, dev);
+
+    /* test if the SPI is set up correctly and the device is responding */
+    if (ng_at86rf2xx_reg_read(dev, NG_AT86RF2XX_REG__PART_NUM) !=
+        NG_AT86RF2XX_PARTNUM) {
+        DEBUG("[ng_at86rf2xx] error: unable to read correct part number\n");
+        return -1;
+    }
+
+    /* reset device to default values and put it into RX state */
+    ng_at86rf2xx_reset(dev);
+    return 0;
+}
+
+void ng_at86rf2xx_reset(ng_at86rf2xx_t *dev)
+{
+    uint8_t tmp;
+#if CPUID_ID_LEN
+    uint8_t cpudid[CPUID_ID_LEN];
+    uint16_t addr_short;
+    uint64_t addr_long;
+#endif
+
+    /* trigger hardware reset */
+    gpio_clear(dev->reset_pin);
+    hwtimer_wait(HWTIMER_TICKS(RESET_DELAY));
+    gpio_set(dev->reset_pin);
+    /* reset options and sequence number */
+    dev->seq_nr = 0;
+    dev->options = 0;
+    /* set short and long address */
+#if CPUID_ID_LEN
+    cpuid_get(cpudid);
+    for (int i = 8; i < CPUID_ID_LEN; i++) {
+        cpudid[i & 0x07] ^= cpudid[i];
+    }
+    /* in cases cpuid len < 8, we just add random memory content */
+    memcpy(&addr_long, cpudid, 8);
```

Comment:
also make sure the least significant bit of the most significant byte is unset (unicast) and the second least significant bit of the most significant byte is set (not globally unique).